### DISCUSSION TITLE
Keycloak extension testing

### DISCUSSION THREAD

**Discussion by koplas** [keycloak/keycloak#30500] [Q&A]
What is the best practice for testing Keycloak extensions that require database access?

**Comment by danielFesenmeyer** [comment]
We are using testcontainers-keycloak for our Keycloak extensions: https://github.com/dasniko/testcontainers-keycloak

Per default, Keycloak uses a h2 in-memory database, but you can easily change this and use a dedicated database container - for the most common databases, integrations are already provided by the testcontainers project: https://java.testcontainers.org/modules/databases/
